toqito/matrices/operator_sinkhorn.py

```diff
@@ -0,0 +1,209 @@
+    `sigma` (:math:`\sigma`) has all of its (single-party) reduced density matrices
+    proportional to the identity, while satisfying
+
+    .. math::
+        \sigma \, = \, F \, \cdot \, \rho \, \cdot \, F^{\dagger}
+
+    In other words, F contains invertible local operations that demonstrate
+    that `rho` and `sigma` are locally equivalent.
+
+    >>> print(sigma)
+    [[ 0.34784186+0.j          0.09278034+0.00551919j -0.02275152+0.05104798j
+       0.20443565-0.12511156j]
+     [ 0.09278034-0.00551919j  0.15215814+0.j          0.1039554 -0.0350973j
+       0.02275152-0.05104798j]
+     [-0.02275152-0.05104798j  0.1039554 +0.0350973j   0.15215814+0.j
+      -0.09278034-0.00551919j]
+     [ 0.20443565+0.12511156j  0.02275152+0.05104798j -0.09278034+0.00551919j
+       0.34784186+0.j        ]]
+
+    >>> print(len(F))
+    2
+
+    F here will be the list of 2 local filtering operators, which are
+
+    >>> print(F[0])
+    [[ 1.33132101+0.00313432j -0.38927801+0.14985847j]
+     [-0.37199021-0.14296461j  1.24635929-0.00300407j]]
+    >>> print(F[1])
+    [[ 1.16317427+0.00765735j -0.16925696+0.09054309j]
+     [-0.18974418-0.09028502j  0.93359381-0.00777829j]]
+
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+    :param rho: input density matrix of a multipartite system
+    :param dim: `None` list containing dimensions of each subsystem.
+                Assumes 2 subsystems with equal dimensions as default.
+    :param tol: `np.sqrt(np.finfo(float).eps)` Convergence tolerance of the iterative Sinkhorn Algorithm.
+                Assumes square root of numpy eps as default.
+    :param max_iterations: Number of iterations after which the solver terminates with a convergence error.
+    :raises: ValueError: if input density matrix is not a square matrix.
+    :raises: ValueError: if the product of dimensions provided/assumed does not match the dimension of density matrix.
+    :raises: RuntimeError: if the sinkhorn algorithm doesnot converge before `max_iterations` iterations.
+    :raises: ValueError: if the density matrix provided is singular (or is not of full rank).
+
+    """
+    # Run checks on the input density matrix
+    rho = np.asarray(rho)
+    if not is_square(rho):
+        raise ValueError("Input 'rho' must be a square matrix.")
+
+    rho = rho.astype(np.complex128)
+
+    dX = rho.shape[0]
+    sdX = int(round(np.sqrt(dX)))
+
+    # set optional argument defaults: dim=sqrt(length(rho)), tol=sqrt(eps).
+    if dim is None:
+        dim = [sdX, sdX]
+
+    # allow the user to enter a single number for dim.
+    if len(dim) == 1:
+        dim.append(dX / dim[0])
+
+        if abs(dim[1] - round(dim[1])) >= (2 * dX * np.finfo(float).eps):
+            raise ValueError(
+                "If dim is of size 1, rho must be square and dim[0] must evenly divide length(rho); "
+                "please provide the dim array containing the dimensions of the subsystems."
```

Comment:
`length` is a MATLAB convention. We should adopt this to be applicable to Python.